### PR TITLE
[IMP] core: handling of errors in migration manager

### PR DESCRIPTION
**Pull request by xmo-odoo**
The handler would interpret *every* attribute error as a missing `migrate` function, so errors in script loading or in the migration script itself might be misattributed to missing migrate functions.

Use `getattr` to properly segregate between the resolution of the `migrate` function and other possible attribute errors.

Also add the script name to the error message, as that's useful information to know what's breaking.

### CODE REVIEW

**Review by aj-fuentes — COMMENTED** [review]
LGTM, minimal suggestion

Extra (out of scope): I think that we could also verify that the function actually have the expected signature before calling it. Example wrong one: `migrate(cr): # forgot version` 

The issue of AttributeError caught here with the misleading subsequent error is annoying to debug. Glad to see this fix :)

**Comment by aj-fuentes** [comment]
> Should probably check that it has at least two positional parameters, and no more than two without defaults?

I think so, yes. We call it always with `cr, version` here, but in some rare cases we have extra parameter with defaults, indeed.

**Review by KangOl — APPROVED** [review]

File: odoo/modules/migration.py [hunk 1]
```diff
@@ -226,18 +226,39 @@ def compare(version):
...
-                        raise
-                    except AttributeError:
-                        _logger.error('module %(addon)s: Each %(stage)s-migration file must have a "migrate(cr, installed_version)" function' % strfmt)
-                    else:
-                        migrate(self.cr, installed_version)
-                    finally:
-                        if mod:
-                            del mod
+                    except ImportError as e:
+                        raise ImportError('module %(addon)s: Unable to load %(stage)s-migration file %(file)s' % dict(strfmt, file=pyfile)) from e
```
Comment (KangOl) [inline odoo/modules/migration.py:232]: The usage of `strfmt` pre-date python3 support. We can use f-strings now.
```suggestion
                        raise ImportError(f'module {addon}: Unable to load {stage}-migration file {pyfile}') from e
```

File: odoo/modules/migration.py [hunk 2]
```diff
@@ -226,18 +226,39 @@ def compare(version):
...
+                        elif p.kind == p.VAR_POSITIONAL:
+                            positionals += 2**8
+                        elif p.kind == p.KEYWORD_ONLY:
+                            required += (p.default is p.empty)
+                        # else p.kind == p.VAR_KEYWORD
+
+                    if positionals < 2:
+                        raise TypeError("module %(addon)s: `migrate` needs at least two positional parameters (cursor and version), %(func)s has %(count)d" % dict(strfmt, func=mod.migrate, count=positionals))
+                    if required > 2:
+                        raise TypeError("module %(addon)s: `migrate` needs at most two required parameters, %(func)s has %(count)d" % dict(strfmt, func=mod.migrate, count=required))
```
Comment (KangOl) [inline odoo/modules/migration.py:261]: We can use pattern matching here.
```python
match list(sig.parameters.values()):
     case [_, _, inspect.Parameter(default=inspect.Parameter.empty), *_]:
          raise TypeError(f"module {addon}: `migrate` needs at most two required parameters")
     case [inspect.Parameter(kind=inspect.Parameter.POSITIONAL_OR_KEYWORD | inspect.Parameter.POSITIONAL_ONLY, default=inspect.Parameter.empty), inspect.Parameter(kind=inspect.Parameter.POSITIONAL_OR_KEYWORD | inspect.Parameter.POSITIONAL_ONLY), *_]:
          pass
     case _:
          raise TypeError(f"module {addon}: `migrate` needs at least two positional parameters (cursor and version)")
```

File: odoo/modules/migration.py (hunk 1 above)
Comment (KangOl) [reply odoo/modules/migration.py:232]: It's a false problem as the current code, even without using f-strings, has still the same issue that this check warn about.

File: odoo/modules/migration.py [hunk 3]
```diff
@@ -226,18 +226,31 @@ def compare(version):
...
+                    try:
+                        sig = inspect.signature(mod.migrate)
+                    except TypeError as e:
+                        raise TypeError("module %(addon)s: `migrate` needs to be a function, got %(migrate)r" % dict(strfmt, migrate=mod.migrate)) from e
+
+                    ps = list(sig.parameters.values())
+                    positional = (inspect.Parameter.POSITIONAL_ONLY, inspect.Parameter.POSITIONAL_OR_KEYWORD)
+                    if not (
+                            len(sig.parameters) == 2
+                        and ps[0].name == 'cr' and ps[0].kind in positional
```
Comment (aj-fuentes) [inline odoo/modules/migration.py:250]: Support also `_cr`.
Example: https://github.com/odoo/upgrade/blob/7917e59312e33dc0b242efc67a74149f0e1383ff/migrations/l10n_be_hr_contract_salary/saas~13.3.1.0/pre-model-contract.py#L5-L6

**Pull request closed without merging**